### ISSUE TITLE
tf.keras in TPU working callbacks call TensorBoard is ERROR

### ISSUE THREAD

**Issue by blackhu** [tensorflow/tensorflow#30528]
**Tensorflow Version:** 1.13.1
**Working Platom:** colab TPU

**Use TensorBoard  Happen Exception：**
“AttributeError: 'TPUFunction' object has no attribute 'fetches'”

**Exception Detail：**
```
/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/callbacks.py in on_epoch_begin(self, epoch, logs)
   1139       # pylint: disable=protected-access
   1140       # add the histogram summary op if it should run this epoch
-> 1141       if self.merged not in self.model._eval_function.fetches:
   1142         self.model._eval_function.fetches.append(self.merged)
   1143         self.model._eval_function.fetch_callbacks[
```

**My Code Fragment：**

```
#TensorBoard
log_dir="logs/fit/"+ datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir=log_dir, histogram_freq=1,write_graph=True, write_images=True)

callbacks_list = [checkpoint,tensorboard_callback]

#Learning
history = tpu_model.fit_generator(generator=train_iter,epochs=100,validation_data=val_iter,callbacks=callbacks_list)
```

**Comment by ravikyram** [comment]
Can you please provide the colab link to reproduce the issue. Thanks!

**Comment by blackhu** [comment]
> 你能否提供colab链接来重现这个问题。谢谢！

https://colab.research.google.com/drive/1VYboO5LMbrcHgTeGmh7Y1Sb4SC_GlwoW#scrollTo=7MgrKf1mvvex

**Comment by ravikyram** [comment]
I am not able to access the Colab link provided. Request you to grant me the access to reproduce the issue.Thanks!

**Comment by blackhu** [comment]
> I am not able to access the Colab link provided. Request you to grant me the access to reproduce the issue.Thanks!

https://colab.research.google.com/drive/1VYboO5LMbrcHgTeGmh7Y1Sb4SC_GlwoW

**Comment by blackhu** [comment]
> I am not able to access the Colab link provided. Request you to grant me the access to reproduce the issue.Thanks!

Full Code:
```
import os
import datetime
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Sequential, Model
from tensorflow.keras import layers
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras import optimizers
from tensorflow.keras.applications.vgg16 import VGG16
from tensorflow.keras.preprocessing import image
from tensorflow.keras.applications.vgg16 import preprocess_input
from tensorflow.keras.callbacks import ModelCheckpoint,TensorBoard
from tensorflow.keras.backend import set_session

from google.colab import drive
from google.colab import files
drive.mount('/dataset')
!mkdir /dataset1
!ln -s /dataset/'My Drive'/newdataset /dataset1

print("Tensorflow Version:",tf.__version__)
print("Tensorflow Keras Version:",tf.keras.__version__)

train_dir = "/dataset1/newdataset/train"
test_dir = "/dataset1/newdataset/test"
!mkdir /dataset1/newdataset/checkpoints
!mkdir /dataset1/newdataset/log
train_pic_gen=ImageDataGenerator(rescale=1./255,rotation_range=20,width_shift_range=0.2,height_shift_range=0.2,shear_range=0.2,zoom_range=0.5,horizontal_flip=True,fill_mode='nearest')
test_pic_gen=ImageDataGenerator(rescale=1./255)
train_iter=train_pic_gen.flow_from_directory(train_dir,(224,224),batch_size=128*2,class_mode='categorical')
val_iter=test_pic_gen.flow_from_directory(test_dir,(224,224),batch_size=128*2,class_mode='categorical')

base_model = VGG16(include_top=False, weights='imagenet', input_shape=(224, 224, 3))
#在vgg16
out = base_model.layers[-1].output
flatten  = layers.Flatten(name="flatten")(out)
fc_1  = layers.Dense(1024, activation='relu')(flatten)
fc_1_droupout = layers.Dropout(0.5)(fc_1)
fc_2 = layers.Dense(512, activation='relu')(fc_1_droupout)
fc_2_droupout = layers.Dropout(0.3)(fc_2)
output = layers.Dense(6, activation='softmax',name="output")(fc_2_droupout)
tuneModel = Model(inputs=base_model.input, outputs = output)
tuneModel.compile(loss='categorical_crossentropy', optimizer= tf.train.RMSPropOptimizer(learning_rate=1e-4) ,metrics=['accuracy'])
#TPU
TPU_WORKER = 'grpc://' + os.environ['COLAB_TPU_ADDR']
tpu_model = tf.contrib.tpu.keras_to_tpu_model(tuneModel,strategy=tf.contrib.tpu.TPUDistributionStrategy(tf.contrib.cluster_resolver.TPUClusterResolver(TPU_WORKER)))
#TensorBoard
log_dir="/dataset1/newdataset/logs/fit/" + datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir=log_dir, histogram_freq=1)
callbacks_list = [tensorboard_callback]

history = tpu_model.fit_generator(generator=train_iter,steps_per_epoch=100,epochs=300,validation_data=val_iter,validation_steps=32,callbacks=callbacks_list)
```

Running After Exception:
```
Drive already mounted at /dataset; to attempt to forcibly remount, call drive.mount("/dataset", force_remount=True).
mkdir: cannot create directory ‘/dataset1’: File exists
ln: failed to create symbolic link '/dataset1/newdataset': File exists
Tensorflow Version: 1.13.1
Tensorflow Keras Version: 2.2.4-tf
mkdir: cannot create directory ‘/dataset1/newdataset/checkpoints’: File exists
mkdir: cannot create directory ‘/dataset1/newdataset/log’: File exists
Found 1411 images belonging to 6 classes.
Found 326 images belonging to 6 classes.
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/resource_variable_ops.py:435: colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Colocations handled automatically by placer.
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/layers/core.py:143: calling dropout (from tensorflow.python.ops.nn_ops) with keep_prob is deprecated and will be removed in a future version.
Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.

WARNING: The TensorFlow contrib module will not be included in TensorFlow 2.0.
For more information, please see:
  * https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md
  * https://github.com/tensorflow/addons
If you depend on functionality not listed there, please file an issue.

INFO:tensorflow:Querying Tensorflow master (grpc://10.26.221.130:8470) for TPU system metadata.
INFO:tensorflow:Found TPU system:
INFO:tensorflow:*** Num TPU Cores: 8
INFO:tensorflow:*** Num TPU Workers: 1
INFO:tensorflow:*** Num TPU Cores Per Worker: 8
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:CPU:0, CPU, -1, 7621372480413121604)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:XLA_CPU:0, XLA_CPU, 17179869184, 16455206848570694857)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:0, TPU, 17179869184, 3055192024204661033)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:1, TPU, 17179869184, 2959337379146398183)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:2, TPU, 17179869184, 8179524105136487244)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:3, TPU, 17179869184, 7965734704452416620)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:4, TPU, 17179869184, 5051508239641363858)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:5, TPU, 17179869184, 7532269395245277437)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:6, TPU, 17179869184, 9200636892927262848)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:7, TPU, 17179869184, 12745496176264895538)
INFO:tensorflow:*** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU_SYSTEM:0, TPU_SYSTEM, 8589934592, 16231103514035382418)
WARNING:tensorflow:tpu_model (from tensorflow.contrib.tpu.python.tpu.keras_support) is experimental and may change or be removed at any time, and without warning.
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-1-996aea90e789> in <module>()
     50 callbacks_list = [tensorboard_callback]
     51 
---> 52 history = tpu_model.fit_generator(generator=train_iter,steps_per_epoch=100,epochs=300,validation_data=val_iter,validation_steps=32,callbacks=callbacks_list)

3 frames
/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/callbacks.py in on_epoch_begin(self, epoch, logs)
   1139       # pylint: disable=protected-access
   1140       # add the histogram summary op if it should run this epoch
-> 1141       if self.merged not in self.model._eval_function.fetches:
   1142         self.model._eval_function.fetches.append(self.merged)
   1143         self.model._eval_function.fetch_callbacks[

AttributeError: 'TPUFunction' object has no attribute 'fetches'

```

**Comment by gowtham-kp** [comment]
@blackhu setting histogram_freq = 1 was making it compute histograms at every epoch. That is the cause of error. Please take a look at this [issue](https://stackoverflow.com/questions/55085661/unable-to-attach-tensorboard-callback-in-tensorflow-keras) where the solution is described and let me know if it helps. Thanks! 
Also please note that this platform is only for bug/performance, feature request or related issues. Please post these kind of issues on Stack overflow as there is a bigger community to help. Thanks!
